src/concretize.rs

```diff
@@ -0,0 +1,282 @@
+//! - [`progress`] makes a state progress (interacting with the buffers)
+//!
+//! And specific implementations of PUT for the different PUTs.
+use crate::agent::{AgentName, TLSVersion};
+use crate::error::Error;
+use crate::io::MessageResult;
+use crate::io::{MemoryStream, Stream};
+use crate::openssl_binding::openssl_version;
+use crate::trace::VecClaimer;
+use crate::wolfssl_binding;
+use crate::{io, openssl_binding};
+use foreign_types_shared::ForeignTypeRef;
+use rustls::msgs::message::OpaqueMessage;
+use security_claims::{deregister_claimer, register_claimer, Claim};
+use serde::{Deserialize, Serialize};
+use std::cell::RefCell;
+use std::io::{Read, Write};
+use std::rc::Rc;
+
+/// Stream, Read, Write traits below are with respect to this content type // [TODO:PUT] how one can make this precise in the type (Without modifing those traits specs?)
+pub type Bts<'a> = &'a [u8];
+
+/// Static configuration for creating a new agent state for the PUT
+pub struct Config {
+    pub tls_version: TLSVersion,
+    /// Whether the agent which holds this descriptor is a server.
+    pub server: bool,
+    ///
+    pub agent_name: AgentName,
+    ///
+    pub claimer: Rc<RefCell<VecClaimer>>,
+}
+
+#[derive(Debug, Copy, Clone, Deserialize, Serialize, Eq, PartialEq)]
+pub enum PUTType {
+    OpenSSL,
+    WolfSSL,
+}
+pub trait PUT: Stream + Drop {
+    /// Create a new agent state for the PUT + set up buffers/BIOs
+    fn new(c: Config) -> Result<Self, Error>
+    where
+        Self: Sized;
+    /// Process incoming buffer, internal progress, can fill in output buffer
+    fn progress(&mut self) -> Result<(), Error>;
+    /// In-place reset of the state
+    fn reset(&mut self) -> Result<(), Error>;
+    /// Register a new claim for agent_name
+    fn register_claimer(&mut self, claimer: Rc<RefCell<VecClaimer>>, agent_name: AgentName);
+    /// Remove all claims in self
+    fn deregister_claimer(&mut self) -> ();
+    /// Change the agent name and the claimer of self
+    fn change_agent_name(&mut self, claimer: Rc<RefCell<VecClaimer>>, agent_name: AgentName);
+    /// Returns a textual representation of the state in which self is
+    fn describe_state(&self) -> &'static str;
+    /// Returns a textual representation of the version of the PUT used by self
+    fn version(&self) -> &'static str;
+    /// Make the PUT used by self determimistic in the future by making its PRNG "deterministic"
+    fn make_deterministic(&self) -> ();
+}
+
+pub fn put_version(put_type: PUTType) -> &'static str {
+    let c = Config {
+        tls_version: TLSVersion::V1_3,
+        server: false,
+        agent_name: AgentName::new(),
+        claimer: Rc::new(RefCell::new(VecClaimer::new())),
+    };
+    let put: Box<dyn PUT> = match { put_type } {
+        OpenSSL => Box::new(OpenSSL::new(c).expect("Failed to create a put instance")),
+        WolfSSL => Box::new(WolfSSL::new(c).expect("Failed to create a put instance")),
+    };
+    put.as_ref().version()
```

Comment:
I think its fine here to construct a whole PUT because `put_version` is only used for debugging.